modules/kvtagger/kvtagger.c

```diff
@@ -0,0 +1,333 @@
+   gsize filename_len = strlen(filename);
+
+   if (filename_len < 3)
+     return NULL;
+
+   gsize diff = filename_len - 3;
+   const gchar *extension = filename + diff;   
+   return extension;
+}
+
+static TagRecordScanner*
+_get_scanner(KVTagger *self)
+{
+  const gchar *type = _get_filename_extension(self->filename);
+  TagRecordScanner *scanner = create_tag_record_scanner_by_type(type);
+
+  if (!scanner)
+    {
+      msg_error("Unknown file extension", evt_tag_str("filename", self->filename));
+      return NULL;
+    }
+
+  tag_record_scanner_set_name_prefix(scanner, self->prefix);
+
+  return scanner;
+}
+
+static gboolean
+_load_tagdb(KVTagger *self)
+{
+  TagRecordScanner *scanner = _get_scanner(self);
+
+  if (!scanner)
+    return FALSE;
+
+  FILE *f = _open_data_file(self->filename);
+  if (!f)
+    {
+      msg_error("Error loading kvtagger database", evt_tag_str("filename", self->filename));
+      tag_record_scanner_free(scanner);
+      return FALSE;
+    }
+
+  gboolean tag_db_loaded = kvtagdb_import(self->tagdb, f, scanner);
+  tag_record_scanner_free(scanner);
+
+  fclose(f);
+  if (!tag_db_loaded)
+    {
+      msg_error("Error while parsing kvtagger database");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_is_initialized(KVTagger *self)
+{
+  return kvtagdb_is_loaded(self->tagdb);
+}
+
+gboolean
+_compile_selector_template(KVTagger *self)
+{
+  GError *error = NULL;
+  if (!self->selector_template_string)
+    {
+      msg_error("No selector set.");
+      return FALSE;
+    }
+
+  if (!log_template_compile(self->selector_template, self->selector_template_string, &error))
+    {
+      msg_error("Failed to compile template",
+                evt_tag_str("template", self->selector_template_string),
+                evt_tag_str("error", error->message));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_first_init(KVTagger *self)
+{
+  if (self->filename == NULL)
+    {
+      msg_error("No database file set.");
+      return FALSE;
+    }
+
+  if (!_compile_selector_template(self))
+    return FALSE;
+
+  if (!_load_tagdb(self))
+    {
+      msg_error("Failed to load the database file.");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+kvtagger_parser_init(LogPipe *s)
+{
+  KVTagger *self = (KVTagger *)s;
+
+  if (_is_initialized(self) || _first_init(self))
+    return log_parser_init_method(s);
+
+  return FALSE;
+}
+
+LogParser *
+kvtagger_parser_new(GlobalConfig *cfg)
+{
+  KVTagger *self = g_new0(KVTagger, 1);
+
+  log_parser_init_instance(&self->super, cfg);
+
+  self->selector_template = log_template_new(cfg, NULL);
+
+  self->super.process = kvtagger_parser_process;
+  self->tagdb = kvtagdb_new();
+
+  self->super.super.clone = kvtagger_parser_clone;
+  self->super.super.free_fn = kvtagger_parser_free;
+  self->super.super.init = kvtagger_parser_init;
+  self->default_selector = NULL;
+  self->selector_template_string = NULL;
+  self->prefix = NULL;
```

Comment:
We don't explicitly initialize members which are already initialized to the same value by `g_new0()` (`0`).